pwiz_tools/Skyline/Model/Import.cs

```diff
@@ -1314,6 +1320,67 @@ private ExTransitionInfo CalcTransitionExplanations(ExTransitionInfo info, strin
                 return info;
             }
 
+            private ProductExp GetPrecursorIsotopeProductExp(TransitionExp transitionExp, double productMz, int? productZ, ExTransitionInfo info)
+            {
+                var precursorAdduct = transitionExp.Precursor.PrecursorAdduct;
+                // If a product charge is specified, it must be the same as the precursor
+                if (productZ.HasValue && productZ.Value != precursorAdduct.AdductCharge)
+                    return null;
+
+                var fullScan = Settings.TransitionSettings.FullScan;
+                if (!fullScan.IsHighResPrecursor)
+                    return null;
+
+                var mods = transitionExp.Precursor.VariableMods;
+                var calc = Settings.GetPrecursorCalc(transitionExp.Precursor.LabelType, mods);
+                var mass = calc.GetPrecursorMass(info.PeptideTarget);
+                var massDist = calc.GetMzDistribution(info.PeptideTarget,
+                    precursorAdduct, fullScan.IsotopeAbundances);
+                var isotopeDist =
+                    IsotopeDistInfo.MakeIsotopeDistInfo(massDist, mass, precursorAdduct, fullScan);
+
+                double minDeltaMz = double.MaxValue;
+                ProductExp productExp = null;
+                for (int i = 0; i < isotopeDist.CountPeaks; i++)
+                {
+                    int massIndex = isotopeDist.PeakIndexToMassIndex(i);
+                    double isotopeMz = isotopeDist.GetMZI(massIndex);
+                    var deltaMz = Math.Abs(productMz - isotopeMz);
+                    if (deltaMz < minDeltaMz && deltaMz <= MzMatchTolerance)
+                    {
+                        productExp = new ProductExp(precursorAdduct, massIndex, info);
+                        minDeltaMz = deltaMz;
+                    }
+                }
+
+                return productExp;
+            }
+
+            /// <summary>
+            /// Finds the closest custom product ion if any to a specified product m/z value.
+            /// </summary>
+            private ProductExp GetCustomIonProductExp(double productMz, ExTransitionInfo info)
+            {
+                double minDeltaMz = double.MaxValue;
+                ProductExp productExp = null;
+                var massType = Settings.TransitionSettings.Prediction.FragmentMassType;
+

```

Comment:
Is it always the case that we aren't provided at least a charge to go with the m/z? That could help disambiguate, when/if available.

Reply:
Right. In proteomics, we are not guaranteed the user supplied a charge. I think the minimum information is sequence, precursor m/z, and product m/z.

Reply:
That's what I figured, but it seems like we ought to consult it when it is provided?

Reply:
Good point. I did that for the precursor charge, but not the custom ion charge. Good catch.